crates/agent/src/controllers/capture.rs

```diff
@@ -84,8 +115,460 @@ impl CaptureStatus {
+    /// This field is cleared as soon as a successful auto-discover is run.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub failure: Option<AutoDiscoverFailure>,
+}
+
+fn interval_schema(_: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+    serde_json::from_value(serde_json::json!({
+        "type": ["string", "null"],
+        "pattern": "^\\d+(s|m|h)$"
+    }))
+    .unwrap()
+}
+
+impl AutoDiscoverStatus {
+    /// Used as the default interval for determining retry backoff when we're
+    /// unable to get the spec for a given connector image/tag.
+    const FALLBACK_INTERVAL: chrono::Duration = chrono::Duration::hours(2);
+
+    async fn try_connector_spec<C: ControlPlane>(
+        model: &models::CaptureDef,
+        control_plane: &mut C,
+    ) -> anyhow::Result<ConnectorSpec> {
+        let models::CaptureEndpoint::Connector(cfg) = &model.endpoint else {
+            anyhow::bail!("only connector endpoints are supported for auto-discovery");
+        };
+        let spec = control_plane
+            .get_connector_spec(cfg.image.clone())
+            .await
+            .context("failed to fetch connector spec")?;
+        if spec.resource_path_pointers.is_empty() {
+            anyhow::bail!("connector has no resource path pointers");
+        }
+        Ok(spec)
+    }
+
+    async fn update<C: ControlPlane>(
+        &mut self,
+        state: &ControllerState,
+        control_plane: &mut C,
+        model: &models::CaptureDef,
+        pending: &mut PendingPublication,
+    ) -> anyhow::Result<NextRun> {
+        // If there's no `connector_tags` row for this capture connector, then we cannot discover.
+        let connector_spec = match Self::try_connector_spec(model, control_plane).await {
+            Ok(s) => s,
+            Err(error) => {
+                tracing::warn!(?error, "auto-discvover failed due to connector spec error");
+                let outcome = AutoDiscoverOutcome::spec_error(
+                    control_plane.current_time(),
+                    &state.catalog_name,
+                    &error,
+                );
+                if let Some(failure) = self.failure.as_mut() {
+                    failure.count += 1;
+                    failure.last_outcome = outcome;
+                } else {
+                    self.failure = Some(AutoDiscoverFailure {
+                        count: 1,
+                        first_ts: control_plane.current_time(),
+                        last_outcome: outcome,
+                    });
+                };
+                return Ok(NextRun::after(
+                    self.next_discover_time(state, Self::FALLBACK_INTERVAL),
+                ));
+            }
+        };
+        let auto_discover_interval = self.interval(connector_spec.auto_discover_interval);
+
+        let next_disco_time = self.next_discover_time(state, auto_discover_interval);
```

Comment:
I saw that, and it does seem like it'd work well for auto-discovers. My thinking had been to hold off on any significant refactors of controllers themselves until we integrate the `automations` crate, because I'd been considering that auto-discovers might actually be separate tasks that are spawned by the main controller. But I'm very low conviction on that point, and happy to take a stab at refactoring this now if you think it's warranted. I'm still on the fence, personally, and so I'm inclined to hold off on a refactor for now, but could be talked into it pretty easily.

Reply:
Oh this discussion comment was intended as setup for the next chunk of work _after_ this PR! No changes please.